authorization/token/service/token_service.go

```diff
@@ -0,0 +1,326 @@
+package service
+
+import (
+	"context"
+	"github.com/dgrijalva/jwt-go"
+	account "github.com/fabric8-services/fabric8-auth/account/repository"
+	"github.com/fabric8-services/fabric8-auth/application/service"
+	"github.com/fabric8-services/fabric8-auth/application/service/base"
+	servicecontext "github.com/fabric8-services/fabric8-auth/application/service/context"
+	tokenPkg "github.com/fabric8-services/fabric8-auth/authorization/token"
+	tokenRepo "github.com/fabric8-services/fabric8-auth/authorization/token/repository"
+	"github.com/fabric8-services/fabric8-auth/errors"
+	"github.com/fabric8-services/fabric8-auth/log"
+	"github.com/fabric8-services/fabric8-auth/token"
+	"github.com/satori/go.uuid"
+
+	"sort"
+	"time"
+)
+
+type TokenServiceConfiguration interface {
+	GetRPTTokenMaxPermissions() int
+}
+
+type tokenServiceImpl struct {
+	base.BaseService
+	config TokenServiceConfiguration
+}
+
+func NewTokenService(context servicecontext.ServiceContext, conf TokenServiceConfiguration) service.TokenService {
+	return &tokenServiceImpl{
+		BaseService: base.NewBaseService(context),
+		config:      conf,
+	}
+}
+
+// Audit verifies an existing token in respect to its privileges for a specified resource.  It starts by validating
+// the status of the token passed in the request, and if that token is currently valid and contains the specified
+// resource, returns the same token.  If the token is invalid or outdated, or doesn't contain the specified resource,
+// then a new token is generated and returned.
+// Returns nil if no new token has been issued, otherwise returns the new token string
+func (s *tokenServiceImpl) Audit(ctx context.Context, identity *account.Identity, tokenString string, resourceID string) (*string, error) {
```

Comment:
Could we verify if the resource exists and return a `BadParameterError`? I looked at the code below and there's no inexpensive operation that determines that.

Reply:
This is reasonable, I've added it to the token service as the first check that is performed.